helper_test.go

```diff
@@ -185,10 +209,72 @@ func echoTCPIdent(tt *tunneltest.TunnelTest, echo, ident string) (string, error)
 	}
 }
 
+func websocketDial(tt *tunneltest.TunnelTest, ident string) (*websocket.Conn, error) {
+	req := tt.Request(ident, nil)
+	if req == nil {
+		return nil, fmt.Errorf("no client found for ident %q", ident)
+	}
+
+	h := http.Header{"Host": {req.Host}}
+	wsurl := fmt.Sprintf("ws://%s", tt.ServerAddr())
+
+	conn, _, err := dialer.Dial(wsurl, h)
+	return conn, err
+}
+
 func sleep() {
 	time.Sleep(time.Duration(rand.Intn(2000)) * time.Millisecond)
 }
 
+func handlerEchoWS(t *testing.T) func(w http.ResponseWriter, r *http.Request) {
+	return handlerEchoSleepWS(t, false)
+}
+
+func handlerLatencyEchoWS(t *testing.T) func(w http.ResponseWriter, r *http.Request) {
+	return handlerEchoSleepWS(t, true)
+}
+
+func handlerEchoSleepWS(t *testing.T, doSleep bool) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		conn, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			t.Errorf("Upgrade error: %s", err)
+			return
+		}
+
+		if doSleep {
+			sleep()
+		}
+
+		for {
+			var msg EchoMessage
+			err := conn.ReadJSON(&msg)
+			if err != nil {
+				t.Errorf("ReadJSON error: %s", err)
+				continue
+			}
+
+			if doSleep {
+				sleep()
+			}
+
+			err = conn.WriteJSON(&msg)
+			if err != nil {
+				t.Errorf("WriteJSON error: %s", err)
```

Comment:
would be better to close the connection in the case of errors, ( setting `msg.Close` to true )